### PR TITLE
Use GN POST to annotate genomic changes

### PR DESCRIPTION
**Pull request by zhx828**
This fixes https://github.com/oncokb/oncokb/issues/3807

This change improves genomic change annotation significantly.

### Before
Average in annotating 159 variants takes 5679ms
master-jmeter

### After
Average in annotating 159 variants takes 482ms
batch-gn-jmeter

### CODE REVIEW

File: core/src/main/java/org/mskcc/cbio/oncokb/util/GenomeNexusUtils.java [hunk 1]
```diff
@@ -215,18 +227,43 @@ private static String getIsoform(Gene gene, ReferenceGenome referenceGenome) {
...
     }
 
-    private static VariantAnnotation getVariantAnnotation(GNVariantAnnotationType type, String query, ReferenceGenome referenceGenome) throws ApiException {
-        VariantAnnotation variantAnnotation = null;
-        if (query != null && type != null) {
+    public static String chromosomeNormalizer(String chromosome) {
+        return chromosome.trim().replace("chr", "").replace("23", "X").replace("24", "Y");
+    }
+
+    private static GenomicLocation convertGenomicLocation(String genomicLocation) {
```
Comment (calvinlu3) [inline core/src/main/java/org/mskcc/cbio/oncokb/util/GenomeNexusUtils.java:234]: Was this method introduced for performance reasons? NotationConverter also has a method to parse genomic location

File: core/src/main/java/org/mskcc/cbio/oncokb/util/GenomeNexusUtils.java [hunk 2]
```diff
@@ -215,18 +227,43 @@ private static String getIsoform(Gene gene, ReferenceGenome referenceGenome) {
...
+        List<VariantAnnotation> variantsAnnotation = null;
+        if (queries != null && type != null) {
             List<String> gnFields = new ArrayList<>();
             gnFields.add("annotation_summary");
             if (type.equals(GNVariantAnnotationType.HGVS_G)) {
-                variantAnnotation = getAnnotationControllerApi(referenceGenome).fetchVariantAnnotationGET(query, MSK_ISOFORM_OVERRIDE, null, gnFields);
+                variantsAnnotation = getAnnotationControllerApi(referenceGenome).fetchVariantAnnotationPOST(queries, MSK_ISOFORM_OVERRIDE, null, gnFields);
             } else {
-                variantAnnotation = getAnnotationControllerApi(referenceGenome).fetchVariantAnnotationByGenomicLocationGET(query, MSK_ISOFORM_OVERRIDE, null, gnFields);
+                variantsAnnotation = getAnnotationControllerApi(referenceGenome).fetchVariantAnnotationByGenomicLocationPOST(queries.stream().map(GenomeNexusUtils::convertGenomicLocation).collect(Collectors.toList()), MSK_ISOFORM_OVERRIDE, null, gnFields);
```
Comment (calvinlu3) [inline core/src/main/java/org/mskcc/cbio/oncokb/util/GenomeNexusUtils.java:263]: If `convertGenomicLocation` returns null, is that supported by the GN POST request? I used the GN swagger to make a POST with an array `[{valid GL object}, null]` and I get a 404 error. We should filter out the null entries?

File: web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/AnnotationsApiController.java [hunk 3]
```diff
@@ -557,23 +612,33 @@ private IndicatorQueryResp getIndicatorQueryFromGenomicLocation(
...
             evidenceTypes,
-                false
+            false
         );
     }
 
+    private Map<String, Alteration> getAlterationsFromGenomeNexus(GNVariantAnnotationType type, ReferenceGenome referenceGenome, List<String> queries) throws ApiException, org.genome_nexus.ApiException {
+        List<String> queriesToGN = new ArrayList<>();
+
+        for (String query : queries) {
```
Comment (calvinlu3) [inline web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/AnnotationsApiController.java:622]: Should we filter out queries that are empty strings at this step. This looks like the entry point to get GN annotation.

File: web/src/main/java/org/mskcc/cbio/oncokb/api/pvt/PrivateGenomeNexusController.java [hunk 4]
```diff
@@ -74,16 +75,20 @@ public ResponseEntity<TranscriptResult> getTranscript(
...
 
         if (body == null) {
             throw new ApiHttpErrorException("The request body is missing.", HttpStatus.BAD_REQUEST);
         } else {
-            for (AnnotateMutationByHGVSgQuery query : body) {
-                GenomeNexusAnnotatedVariantInfo resp = GenomeNexusUtils.getAnnotatedVariantFromGenomeNexus(GNVariantAnnotationType.HGVS_G, query.getHgvsg(), query.getReferenceGenome());
-                result.add(resp);
+            ReferenceGenome referenceGenome = null;
+            if (!body.isEmpty()) {
+                referenceGenome = body.get(0).getReferenceGenome();
```
Comment (calvinlu3) [inline web/src/main/java/org/mskcc/cbio/oncokb/api/pvt/PrivateGenomeNexusController.java:86]: Its not guaranteed that the reference genome is the same for the entire List<AnnotateMutationByHGVSgQuery>, but we're using the first ReferenceGenome here.

If we were to support a POST query with potentially different ReferenceGenome, then we probably have to partition the list into two, get GN annotation, then merge together?

File: core/src/test/java/org/mskcc/cbio/oncokb/util/IndicatorUtilsTest.java
Comment (calvinlu3) [inline core/src/test/java/org/mskcc/cbio/oncokb/util/IndicatorUtilsTest.java:1]: Potentially some additional tests cases for GNVariantAnnotationType genomic location and some edge cases would make this test suite more complete.

**Review by calvinlu3 — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/mskcc/cbio/oncokb/util/GenomeNexusUtils.java (hunk 1 above)
Comment (zhx828) [reply core/src/main/java/org/mskcc/cbio/oncokb/util/GenomeNexusUtils.java:234]: Issue seems to be util method can't use the NotationConverter class.

File: core/src/main/java/org/mskcc/cbio/oncokb/util/GenomeNexusUtils.java (hunk 2 above)
Comment (zhx828) [reply core/src/main/java/org/mskcc/cbio/oncokb/util/GenomeNexusUtils.java:263]: My thinking is we should not. We've been trying to persist the query location in the list. WI'm thinking to return null for the null input.

File: web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/AnnotationsApiController.java (hunk 3 above)
Comment (zhx828) [reply web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/AnnotationsApiController.java:622]: `genomicLocationShouldBeAnnotated` will filter out the empty query.

File: core/src/main/java/org/mskcc/cbio/oncokb/util/GenomeNexusUtils.java (hunk 2 above)
Comment (calvinlu3) [reply core/src/main/java/org/mskcc/cbio/oncokb/util/GenomeNexusUtils.java:263]: Gotcha, persisting the index is valid. However, do you know if passing in a null value in the POST body array will throw an exception for the entire request even if there are valid queries to annotate?

Comment (zhx828) [reply core/src/main/java/org/mskcc/cbio/oncokb/util/GenomeNexusUtils.java:263]: null in request body or in response?

**Review by calvinlu3 — APPROVED** [review]
LGTM
